src/coreComponents/physicsSolvers/fluidFlow/wells/PerforationFluxKernels.hpp

```diff
@@ -0,0 +1,881 @@
+                        &dVisc[CP_Deriv::dC],
+                        Deriv::dC );
+
+        // relative permeability
+        real64 const resRelPerm = m_resPhaseRelPerm[er][esr][ei][0][ip];
+        real64 dRelPerm[CP_Deriv::nDer]{};
+        for( integer jc = 0; jc < CP_Deriv::nDer; ++jc )
+        {
+          dRelPerm[jc]=0;
+        }
+        for( integer jp = 0; jp < NP; ++jp )
+        {
+          real64 const dResRelPerm_dS = m_dResPhaseRelPerm_dPhaseVolFrac[er][esr][ei][0][ip][jp];
+          dRelPerm[CP_Deriv::dP] += dResRelPerm_dS * m_dResPhaseVolFrac[er][esr][ei][jp][Deriv::dP];
+          if constexpr ( IS_THERMAL )
+          {
+            dRelPerm[CP_Deriv::dT] += dResRelPerm_dS * m_dResPhaseVolFrac[er][esr][ei][jp][Deriv::dT];
+          }
+          for( integer jc = 0; jc < NC; ++jc )
+          {
+            dRelPerm[CP_Deriv::dC+jc] += dResRelPerm_dS * m_dResPhaseVolFrac[er][esr][ei][jp][Deriv::dC+jc];
+          }
+        }
+
+        // compute the reservoir phase mobility, including phase density
+        real64 const resPhaseMob = resDens * resRelPerm / resVisc;
+
+        // Handles all dependencies
+        for( integer jc = 0; jc < CP_Deriv::nDer; ++jc )
+        {
+          dMob[jc] = dRelPerm[jc] * resDens / resVisc
+                     + resPhaseMob * (dDens[jc] / resDens - dVisc[jc] / resVisc);
+        }
+        // compute the phase flux and derivatives using upstream cell mobility
+        flux = resPhaseMob * potDiff;
+        // Handles all dependencies
+        for( integer jc = 0; jc < CP_Deriv::nDer; ++jc )
+        {
+          dFlux[TAG::RES][jc]  = dMob[jc] * potDiff + resPhaseMob * dPotDiff[TAG::RES][jc];
+          dFlux[TAG::WELL][jc] = resPhaseMob * dPotDiff[TAG::WELL][jc];
+        }
+
+        // increment component fluxes
+        for( integer ic = 0; ic < NC; ++ic )
+        {
+          // Note this needs to be uncommented out
+          m_compPerfRate[iperf][ic] += flux *  m_resPhaseCompFrac[er][esr][ei][0][ip][ic];
+          dCompFrac[CP_Deriv::dP] = m_dResPhaseCompFrac[er][esr][ei][0][ip][ic][Deriv::dP];
+          if constexpr (IS_THERMAL)
+          {
+            dCompFrac[CP_Deriv::dT] = m_dResPhaseCompFrac[er][esr][ei][0][ip][ic][Deriv::dT];
+          }
+
+          applyChainRule( NC,
+                          m_dResCompFrac_dCompDens[er][esr][ei],
+                          m_dResPhaseCompFrac[er][esr][ei][0][ip][ic],
+                          &dCompFrac[CP_Deriv::dC],
+                          Deriv::dC );
+
+          for( integer jc = 0; jc < CP_Deriv::nDer; ++jc )
+          {
+            m_dCompPerfRate[iperf][TAG::RES][ic][jc]  += dFlux[TAG::RES][jc] *  m_resPhaseCompFrac[er][esr][ei][0][ip][ic];
+            m_dCompPerfRate[iperf][TAG::RES][ic][jc]  += flux * dCompFrac[jc];
+            m_dCompPerfRate[iperf][TAG::WELL][ic][jc] += dFlux[TAG::WELL][jc] *  m_resPhaseCompFrac[er][esr][ei][0][ip][ic];
+          }
+        }
+        if constexpr ( IS_THERMAL )
```

Comment:
you don't need these if `constexpr` nor the IS_THERMAL template parameter.

Reply:
No, the code is more readable this way